README
======================
ResourceSync Simulator
======================

The ResourceSync Simulator simulates a `ResourceSync
<http://www.openarchives.org/rs>`_ Source, which is a server that hosts resources subject to synchronization.

Any ResourceSync-compliant client can be used to synchronize a Destination with the simulated Source. This version of the simulator is written against v0.9.1 of the `ResourceSync specification
<http://www.openarchives.org/rs/0.9.1>`_ and has been tested with v0.9.3 of our own `ResourceSync client and library reference implementation
<https://github.com/resync/resync>`_.

Installation
------------

Make sure Python 2.7.2 or later is running on your system::

    python --version

**Automatic installation**::

    easy_install resync-simulator

``rsync-simulator`` is listed in `PyPI
<http://pypi.python.org/pypi/resync-simulator>`_ and can be installed with ``pip`` or ``easy_install``. Doing an ``easy_install`` should also install the dependencies ``resync`` and ``tornado`` if they are not already on your system.

**Manual installation from github**:

Install the `resync
<https://github.com/resync/resync>`_ library code and the `Tornado
<http://www.tornadoweb.org/>`_ web server::

    sudo easy_install resync
    sudo easy_install tornado
    
Get the ResourceSync Simulator from `Github
<http://www.github.com/resync/simulator>`_::

    git clone git://github.com/resync/simulator.git

Quick Start
-----------

Run the source simulator (with the default configuration in ``./config/default.yaml``)::

    ./resync-simulator

Access from a web browser as http://localhost:8888/

Terminate the source simulator::

    CTRL-C

How to define parameterized use cases
-------------------------------------

Parameterized Use Cases can be defined by creating a `YAML
<http://www.yaml.org/>`_ configuration file (e.g., ``simulation1.yaml`` and defining a set of parameters::

    source:
        name: ResourceSync Simulator
        number_of_resources: 1000
        change_delay: 2
        event_types: [create, update, delete]
        average_payload: 1000
        max_events: -1
        stats_interval: 10
        
Additional **resource_list_builder** and **change memory** implementations can be attached for simulation purposes. For instance, the following configuration attaches a change memory implemented by the DynamicChangeList class::

    resource_list_builder:
        class: DynamicResourceListBuilder
        uri_path: resourcelist.xml

    changememory:
        class: DynamicChangeList
        uri_path: changelist.xml
        max_changes: 1000
            
See the examples in the **./config** directory for further details.

See also
--------

`ResourceSync library
<http://github.com/resync/resync>`_

Author and Contributors
-----------------------

Author:
`Bernhard Haslhofer
<https://github.com/behas>`_

Contributors:
`Simeon Warner
<https://github.com/zimeon>`_


CHANGELOG
2013-11-13
    RELEASE 0.8
    
    * change package name to resync_simulator
    * add source_description_uri option as source config
    
2013-09-06
    RELEASE 0.7
    
    * updated for v0.9.1 ResourceSync specification
    * updated to use resync v0.9.3 library

2013-05-14
    RELEASE 0.6

    * implemented simulator with resync v0.6.2 library
    * dead code parts cleanup

2013-05-07
    RELEASE 0.5

    * removed experimental code
    * removed dynamic changememory based on changeid
    * removed static changememory builder (which needs to be re-written)
    * tested with current resync client master


2012-08-24
    RELEASE 0.4
    
    * implemented static changememory
    * added max events restriction to changememory
    * implemented changesets (client side)
    * implemented dump (client side)
    * added static sitemap creation option to simulator
    * unified simulator and client sitemap creation

2012-06-13
    RELEASE 0.3
    
    * unified client and source inventory implementation
    * wired dynamic inventory with changememory
    * implemented Dynamic ChangeSet according to draft specification
    * add resync-client for baseline synchronization
    * added XMPP publisher implementation
    * Generalized resource.py for re-use on client and server side
    * Added change memory with paging implementation

2012-05-08
    RELEASE 0.2
    
    * Implemented command line publisher
    * Implemented dynamic digest
    * Implemented dynamic sitemap inventory
    * Implemented reflective loading of simulation components
    * Modularized architecture: source - inventory - change memory - publisher
    * Implemented command line eventlog
    * Extracted change model into separate module
    * Merged simulator and inventory logic into "source"
    * Cleaned up simulation configuration; switched to YAML

2012-04-04
    RELEASE 0.1
    
    * Initial release supporting basic simulation, observer interface, sitemap
    generation, and a simple Web interface.

resync/source.py
#!/usr/bin/env python
# encoding: utf-8
"""
source.py: A source holds a set of resources and changes over time.

Resources are internally stored by their basename (e.g., 1) for memory
efficiency reasons.

Created by Bernhard Haslhofer on 2012-04-24.
Copyright 2012, ResourceSync.org. All rights reserved.
"""

import re
import os
import random
import pprint
import logging
import time
import shutil

from apscheduler.scheduler import Scheduler

from resync.observer import Observable
from resync.resource_change import ResourceChange
from resync.resource import Resource
from resync.digest import compute_md5_for_string
from resync.inventory import Inventory
from resync.sitemap import Sitemap, Mapper

#### Source-specific capability implementations ####

class DynamicInventoryBuilder(object):
    """Generates an inventory snapshot from a source"""
    
    def __init__(self, source, config):
        self.source = source
        self.config = config
        self.logger = logging.getLogger('inventory_builder')
        
    def bootstrap(self):
        """Bootstrapping procedures implemented in subclasses"""
        pass
    
    @property
    def path(self):
        """The inventory path (from the config file)"""
        return self.config['uri_path']

    @property
    def uri(self):
        """The inventory URI (e.g., http://localhost:8080/sitemap.xml)"""
        return self.source.base_uri + "/" + self.path
    
    def generate(self):
        """Generates an inventory (snapshot from the source)"""
        then = time.time()
        capabilities = {}
        if self.source.has_changememory:
            next_changeset = self.source.changememory.next_changeset_uri()
            capabilities[next_changeset] = {"rel": "next http://www.openarchives.org/rs/changeset"}
        inventory = Inventory(resources=self.source.resources,
                              capabilities=capabilities)
        now = time.time()
        self.logger.info("Generated inventory: %f" % (now-then))
        return inventory
        
class StaticInventoryBuilder(DynamicInventoryBuilder):
    """Periodically writes an inventory to the file system"""
    
    def __init__(self, source, config):
        super(StaticInventoryBuilder, self).__init__(source, config)
                                
    def bootstrap(self):
        """Bootstraps the static inventory writer background job"""
        self.rm_sitemap_files(Source.STATIC_FILE_PATH)
        self.write_static_inventory()
        logging.basicConfig()
        interval = self.config['interval']
        sched = Scheduler()
        sched.start()
        sched.add_interval_job(self.write_static_inventory,
                                seconds=interval)
        
    def write_static_inventory(self):
        """Writes the inventory to the filesystem"""
        # Generate sitemap in temp directory
        then = time.time()
        self.ensure_temp_dir(Source.TEMP_FILE_PATH)
        inventory = self.generate()
        basename = Source.TEMP_FILE_PATH + "/sitemap.xml"
        s=Sitemap()
        s.max_sitemap_entries=self.config['max_sitemap_entries']
        s.mapper=Mapper([self.source.base_uri, Source.TEMP_FILE_PATH])
        s.write(inventory, basename)
        # Delete old sitemap files; move the new ones; delete the temp dir
        self.rm_sitemap_files(Source.STATIC_FILE_PATH)
        self.mv_sitemap_files(Source.TEMP_FILE_PATH, Source.STATIC_FILE_PATH)
        shutil.rmtree(Source.TEMP_FILE_PATH)
        now = time.time()
        # Log Sitemap create start event
        sitemap_size = self.compute_sitemap_size(Source.STATIC_FILE_PATH)
        log_data = {'time': (now-then), 
                    'no_resources': self.source.resource_count}
        self.logger.info("Wrote static sitemap inventory. %s" % log_data)
        sm_write_end = ResourceChange(
                resource = ResourceChange(self.uri, 
                                size=sitemap_size,
                                timestamp=then),
                                changetype = "UPDATED")
        self.source.notify_observers(sm_write_end)
        
    def ensure_temp_dir(self, temp_dir):
        """Create temp directory if it doesn't exist; removes existing one"""
        if os.path.exists(temp_dir):
            shutil.rmtree(temp_dir)
            os.makedirs(temp_dir)
        else:
            os.makedirs(temp_dir)
    
    def ls_sitemap_files(self, directory):
        """Returns the list of sitemaps in a directory"""
        p = re.compile('sitemap\d*\.xml')
        filelist = [ f for f in os.listdir(directory) if p.match(f) ]
        return filelist
    
    def rm_sitemap_files(self, directory):
        """Deletes sitemap files (from previous runs)"""
        filelist = self.ls_sitemap_files(directory)
        if len(filelist) > 0:
            self.logger.debug("*** Cleaning up %d sitemap files ***" % 
                                                                len(filelist))
            for f in filelist:
                filepath = directory + "/" + f
                os.remove(filepath)
    
    def mv_sitemap_files(self, src_directory, dst_directory):
        """Moves sitemaps from src to dst directory"""
        filelist = self.ls_sitemap_files(src_directory)
        if len(filelist) > 0:
            self.logger.debug("*** Moving %d sitemap files ***" % 
                                                                len(filelist))
            for f in filelist:
                filepath = src_directory + "/" + f
                shutil.move(filepath, dst_directory)
    
    def compute_sitemap_size(self, directory):
        """Computes the size of all sitemap files in a given directory"""
        return sum([os.stat(directory + "/" + f).st_size 
                        for f in self.ls_sitemap_files(directory)])
    
#### Source Simulator ####

class Source(Observable):
    """A source contains a list of resources and changes over time"""
    
    RESOURCE_PATH = "/resources"
    STATIC_FILE_PATH = os.path.join(os.path.dirname(__file__), "static")
    TEMP_FILE_PATH = os.path.join(os.path.dirname(__file__), "temp")
    
    def __init__(self, config, hostname, port):
        """Initalize the source"""
        super(Source, self).__init__()
        self.logger = logging.getLogger('source')
        self.config = config
        self.logger.info("Source config: %s " % self.config)
        self.hostname = hostname
        self.port = port
        self.max_res_id = 1
        self._repository = {} # {basename, {timestamp, size}}
        self.inventory_builder = None # The inventory builder implementation
        self.changememory = None # The change memory implementation
        self.no_events = 0
    
    ##### Source capabilities #####
    
    def add_inventory_builder(self, inventory_builder):
        """Adds an inventory builder implementation"""
        self.inventory_builder = inventory_builder
        
    @property
    def has_inventory_builder(self):
        """Returns True in the Source has an inventory builder"""
        return bool(self.inventory_builder is not None)        
    
    def add_changememory(self, changememory):
        """Adds a changememory implementation"""
        self.changememory = changememory
        
    @property
    def has_changememory(self):
        """Returns True if a source maintains a change memory"""
        return bool(self.changememory is not None)
    
    ##### Bootstrap Source ######

    def bootstrap(self):
        """Bootstrap the source with a set of resources"""
        self.logger.info("Bootstrapping source...")
        for i in range(self.config['number_of_resources']):
            change = self._create_resource(notify_observers = False)
            self.logger.debug("Event: %s" % repr(change))
        if self.has_changememory: self.changememory.bootstrap()
        if self.has_inventory_builder: self.inventory_builder.bootstrap()
        self._log_stats()
    
    ##### Source data accessors #####
    
    @property
    def base_uri(self):
        """Returns the base URI of the source (e.g., http://localhost:8888)"""
        return "http://" + self.hostname + ":" + str(self.port)

    @property
    def resource_count(self):
        """The number of resources in the source's repository"""
        return len(self._repository)
    
    @property
    def resources(self):
        """Iterates over resources and yields resource objects"""
        for basename in self._repository.keys():
            resource = self.resource(basename)
            if resource is None:
                self.logger.error("Cannot create resource %s " % basename + \
                      "because source object has been deleted.")
            else:
                yield resource
    
    @property
    def random_resource(self):
        """Returns a single random resource"""
        rand_res = self.random_resources()
        if len(rand_res) == 1:
            return rand_res[0]
        else:
            return None
    
    def resource(self, basename):
        """Creates and returns a resource object from internal resource
        repository. Repositoy values are copied into the object."""
        if not self._repository.has_key(basename): return None
        uri = self.base_uri + Source.RESOURCE_PATH + "/" + basename
        timestamp = self._repository[basename]['timestamp']
        size = self._repository[basename]['size']
        md5 = compute_md5_for_string(self.resource_payload(basename, size))
        return Resource(uri = uri, timestamp = timestamp, size = size,
                        md5 = md5)
    
    def resource_payload(self, basename, size = None):
        """Generates dummy payload by repeating res_id x size times"""
        if size == None: size = self._repository[basename]['size']
        no_repetitions = size / len(basename)
        content = "".join([basename for x in range(no_repetitions)])
        no_fill_chars = size % len(basename)
        fillchars = "".join(["x" for x in range(no_fill_chars)])
        return content + fillchars
    
    def random_resources(self, number = 1):
        "Return a random set of resources, at most all resources"
        if number > len(self._repository):
            number = len(self._repository)
        rand_basenames = random.sample(self._repository.keys(), number)
        return [self.resource(basename) for basename in rand_basenames]
    
    def simulate_changes(self):
        """Simulate changing resources in the source"""
        self.logger.info("Starting simulation...")
        sleep_time = self.config['change_delay']
        while self.no_events != self.config['max_events']:
            time.sleep(sleep_time)
            event_type = random.choice(self.config['event_types'])
            if event_type == "create":
                self._create_resource()
            elif event_type == "update" or event_type == "delete":
                if len(self._repository.keys()) > 0:
                    basename = random.sample(self._repository.keys(), 1)[0]
                else:
                    basename = None
                if basename is None: 
                    self.no_events = self.no_events + 1                    
                    continue
                if event_type == "update":
                    self._update_resource(basename)
                elif event_type == "delete":
                    self._delete_resource(basename)

            else:
                self.logger.error("Event type %s is not supported" 
                                                                % event_type)
            self.no_events = self.no_events + 1
            if self.no_events%self.config['stats_interval'] == 0:
                self._log_stats()

        self.logger.info("Finished change simulation")
    
    # Private Methods
    
    def _create_resource(self, basename = None, notify_observers = True):
        """Create a new resource, add it to the source, notify observers."""
        if basename == None:
            basename = str(self.max_res_id)
            self.max_res_id += 1
        timestamp = time.time()
        size = random.randint(0, self.config['average_payload'])
        self._repository[basename] = {'timestamp': timestamp, 'size': size}
        change = ResourceChange(resource = self.resource(basename),
                                changetype = "CREATED")
        if notify_observers:
            self.notify_observers(change)
            self.logger.debug("Event: %s" % repr(change))
        return change
        
    def _update_resource(self, basename):
        """Update a resource, notify observers."""
        self._delete_resource(basename, notify_observers = False)
        self._create_resource(basename, notify_observers = False)
        change = ResourceChange(resource = self.resource(basename),
                                changetype = "UPDATED")
        self.notify_observers(change)
        self.logger.debug("Event: %s" % repr(change))

    def _delete_resource(self, basename, notify_observers = True):
        """Delete a given resource, notify observers."""
        res = self.resource(basename)
        del self._repository[basename]
        res.timestamp = time.time()
        if notify_observers:
            change = ResourceChange(resource = res, changetype = "DELETED")
            self.notify_observers(change)
            self.logger.debug("Event: %s" % repr(change))
    
    def _log_stats(self):
        """Log current source statistics"""
        stats = {
            'no_resources': self.resource_count,
            'no_events': self.no_events
        }
        self.logger.info("Source stats: %s" % stats)
    
    def __str__(self):
        """Prints out the source's resources"""
        return pprint.pformat(self._repository)
